relay-server/src/utils/rate_limits.rs

```diff
@@ -352,38 +386,69 @@ where
+
         Ok((enforcement, rate_limits))
     }
 
     fn execute(
         &mut self,
         summary: &EnvelopeSummary,
         scoping: &Scoping,
+        early_enforcement: &Enforcement,
     ) -> Result<(Enforcement, RateLimits), E> {
         let mut rate_limits = RateLimits::new();
         let mut enforcement = Enforcement::default();
 
         if let Some(category) = summary.event_category {
-            let event_limits = (self.check)(scoping.item(category), 1)?;
-            let longest = event_limits.longest();
-            enforcement.event = CategoryLimit::new(category, 1, longest);
-
-            // Record the same reason for attachments, if there are any.
-            enforcement.attachments = CategoryLimit::new(
-                DataCategory::Attachment,
-                summary.attachment_quantity,
-                longest,
-            );
+            if early_enforcement.event.is_active() {
+                enforcement.event = early_enforcement.event.clone();
+                // No need to set a RateLimit for this, we only need to report new rate
+                // limits as those are the ones that need to be sent to the ProjectCache.
+            } else {
+                let event_limits = (self.check)(scoping.item(category), 1)?;
+                let longest = event_limits.longest();
+                enforcement.event = CategoryLimit::new(category, 1, longest);
+
+                // Record the same reason for attachments, if there are any.
+                enforcement.attachments = CategoryLimit::new(
+                    DataCategory::Attachment,
+                    summary.attachment_quantity,
+                    longest,
+                );
+
+                rate_limits.merge(event_limits);
+            }
 
-            rate_limits.merge(event_limits);
+            // Handle transactions specially, they have processing quota too.
+            if category == DataCategory::Transaction {
+                if early_enforcement.transaction_processed.is_active() {
+                    enforcement.transaction_processed =
+                        early_enforcement.transaction_processed.clone()
+                } else {
+                    let item_scoping = scoping.item(DataCategory::TransactionProcessed);
+                    let limits = (self.check)(item_scoping, 1)?;
+                    enforcement.transaction_processed =
+                        CategoryLimit::new(DataCategory::TransactionProcessed, 1, limits.longest());
+                    rate_limits.merge(limits);
+
+                    // If only one of the rate limits applied we omit both of them.  If there is
+                    // a rate limit the endpoint will return 429 but we need the client to keep
+                    // sending transactions unless both quotas limits were exceeded.
+                    if rate_limits.iter().count() == 1 {
+                        rate_limits = RateLimits::new();
+                    }
```

Comment:
Is there a scenario where rate limits apply for indexed but not for processed transactions? So far I assumed that an active rate limit on processed transactions implies an active rate limit on indexed transactions.